### PR TITLE
Hsd8 752 google forms media

### PR DESCRIPTION
**Pull request by pookmish**
# READY FOR REVIEW

# Summary
- Added google forms media source

# Need Review By (Date)
- 4/8

# Urgency
- medium

# Steps to Test
1. checkout this branch.
1. checkout https://github.com/SU-SWS/stanford_profile/pull/169
1. `drush cim -y`
1. create a google form at https://docs.google.com/forms/u/0/ without a file upload field
1. go to a wysiwyg and go to add a media item
1. paste the link to the google form
1. validate the form is embedded in an iframe.

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

### CODE REVIEW

**Comment by pookmish** [comment]
poke

**Comment by sherakama** [comment]
Thanks for the poke. I'm doing my best to get to it but I am backlogged at the moment.

File: src/Plugin/Validation/Constraint/GoogleFormsConstraint.php [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ *   id = "google_forms",
+ *   label = @Translation("Google Forms", context = "Validation"),
+ *   type = "string"
+ * )
+ */
+class GoogleFormsConstraint extends Constraint {
+
+  public $invalidString = 'The given data does not contain a valid url';
+
+  public $invalidUrl = 'The given URL is not a google forms url.';
```
Comment (sherakama) [inline src/Plugin/Validation/Constraint/GoogleFormsConstraint.php:20]: Do these get translated somewhere?

File: src/Plugin/Field/FieldFormatter/GoogleFormFormatter.php [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  public function viewElements(FieldItemListInterface $items, $langcode) {
+    $elements = [];
+    /** @var \Drupal\Core\Field\Plugin\Field\FieldType\StringItem $item */
+    foreach ($items as $item) {
+      $url = $item->getValue()['value'];
+      $elements[] = [
+        '#type' => 'html_tag',
+        '#tag' => 'iframe',
+        '#value' => $this->t('Loading'),
+        '#attributes' => ['src' => $url, 'class' => ['google-form']],
```
Comment (sherakama) [inline src/Plugin/Field/FieldFormatter/GoogleFormFormatter.php:49]: iframes need an accessible title. https://dequeuniversity.com/tips/provide-iframe-titles

Comment (pookmish) [reply src/Plugin/Field/FieldFormatter/GoogleFormFormatter.php:49]: Good catch

File: src/Plugin/Validation/Constraint/GoogleFormsConstraint.php (hunk 1 above)
Comment (pookmish) [reply src/Plugin/Validation/Constraint/GoogleFormsConstraint.php:20]: yeah they do https://git.drupalcode.org/project/drupal/-/blob/8.8.x/core/lib/Drupal/Core/TypedData/Validation/ExecutionContext.php#L166

File: src/Plugin/media/Source/GoogleForm.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  /**
+   * {@inheritDoc}
+   */
+  public function getMetadata(MediaInterface $media, $attribute_name) {
+    switch ($attribute_name) {
+      case self::METADATA_ATTRIBUTE_ID:
+        return $this->getId($media);
+
+    }
+    return parent::getMetadata($media, $attribute_name);
```
Comment (sherakama) [inline src/Plugin/media/Source/GoogleForm.php:47]: Why the switch statement and no default?

Comment (pookmish) [reply src/Plugin/media/Source/GoogleForm.php:47]: i've seen the switch as a normal convention on these methods within other media types. for the quick addition of other metadata. guess the bottom line there acts as the `default` case

Comment (sherakama) [reply src/Plugin/media/Source/GoogleForm.php:47]: Not your normal style but not unreadable or a performance hit. Meh.

File: src/Plugin/Field/FieldFormatter/GoogleFormFormatter.php (hunk 2 above)
Comment (sherakama) [reply src/Plugin/Field/FieldFormatter/GoogleFormFormatter.php]: Thanks.

File: src/Plugin/Validation/Constraint/GoogleFormsConstraint.php (hunk 1 above)
Comment (sherakama) [reply src/Plugin/Validation/Constraint/GoogleFormsConstraint.php:20]: Thanks.

**Comment by sherakama** [comment]
Works as designed. I submitted a question to the sws-d8core slack channel asking if we should add a paragraph type for this as embedding is slightly obfuscated through the embed media button in the editor.

**Pull request merged**